ehr/resources/pipeline/kinship/populateKinship.r

```diff
@@ -35,69 +36,35 @@ allPed$Species <- as.character(allPed$Species)
-# it is similar to add.Inds from kinship; however, we retain gender
-addMissing <- function(ped, species){
-    if(ncol(ped)<4)stop("pedigree should have at least 4 columns")
-    head <- names(ped)
-
-    nsires <- match(ped[,3],ped[,1])# [Quoc] change ped,2 to ped,3
-    nsires <- as.character(unique(ped[is.na(nsires),3]))
-    nsires <- nsires[!is.na(nsires)]
-    if(length(nsires)){
-        ped <- rbind(ped, data.frame(Id=nsires, Dam=rep(NA, length(nsires)), Sire=rep(NA, length(nsires)), Gender=rep(1, length(nsires)), Species=rep(species, length(nsires))));
-    }
-
-    ndams <- match(ped[,2],ped[,1])# [Quoc] change ped,3 to ped,2
-    ndams <- as.character(unique(ped[is.na(ndams),2]))
-    ndams <- ndams[!is.na(ndams)];
-
-    if(length(ndams)){
-        ped <- rbind(ped,data.frame(Id=ndams, Dam=rep(NA, length(ndams)), Sire=rep(NA, length(ndams)), Gender=rep(2, length(ndams)), Species=rep(species, length(ndams))));
-    }
-
-    names(ped) <- head
-    return(ped)
-}
-
 # In order to reduce the max matrix size, calculate famids using makefamid, then analyze each group separately
 # It resizes the biggest matrix from 12000^2 to 8200^2 thus reduces the memory used by half
 newRecords=NULL
 for (species in unique(allPed$Species)){
     print(paste0('processing species: ', species))
-    recordsForSpecies <- allPed[allPed$Species == species,]
-    print(paste0('total subjects: ', nrow(recordsForSpecies)))
-
-    recordsForSpecies <- addMissing(recordsForSpecies, species)
-    gc()
-    print(paste0('total subjects after adding missing: ', nrow(recordsForSpecies)))
-
-    fami=makefamid(id=recordsForSpecies$Id,father.id=recordsForSpecies$Sire,mother.id=recordsForSpecies$Dam)
-    famid=unique(fami)
-    famid=famid[famid!=0]
-    gc()
-
-    print(paste0('total families: ', length(famid)))
-    for (fam.no in famid){
-        familytemp=recordsForSpecies[fami==fam.no,]
-        print(paste0('family size: ', nrow(familytemp)))
-
-        temp.kin=kinship(familytemp$Id,familytemp$Sire,familytemp$Dam)
-        rownames(temp.kin) <- colnames(temp.kin) #add rownames to make matrix symmetric, which is required downstream
-        sparse.kin=as(temp.kin,"dsTMatrix") #change kinship matrix to symmetric triplet sparse matrix
-
-        # convert to a sparse matrix usng S4 object from Matrix library
-        temp.tri=data.frame(Id=colnames(temp.kin)[sparse.kin@i+1],Id2=colnames(temp.kin)[sparse.kin@j+1],coefficient=sparse.kin@x,stringsAsFactors=FALSE)
-        newRecords=rbind(newRecords,temp.tri)
-    }
+    allRecordsForSpecies <- allPed[allPed$Species == species,]
+
+    # Add missing parents for accurate kinship calculations
+    fixedRecords <- with(allRecordsForSpecies, fixParents(id = Id, dadid = Sire, momid = Dam, sex = Gender))
+
+    # Kinship is expecting records to be sorted IAW it's own pedigree function
+    recordsForSpecies <- with(fixedRecords, pedigree(id=id,dadid=dadid,momid=momid,sex=sex,missid=0))
```

Comment:
The original rationale in splitting the data by kinship's makefamid was to reduce the data volume and calculation time. how big a dataset have you used here? this was written in probably 2010, so it's possible either modern servers and/or improvements to kinship() mean that's no longer necessary. It's certainly probably better if we can compute everything in one call per species, so that we dont get false-negatives from relatives being split into different families by makefamid. do oyu have the logs from a previously successful ONPRC kinship run, to see how many inferred families there were, and how big those families were?

Reply:
This actually considerably speeds up the calculations, at least on my laptop and did not run into issues with resources. I suspect any servers will have more resources than my laptop. I tried out a pretty complex colony dataset with about 40k animals and wide variation of family sizes.

The part that can have performance issues is the optional validation I added that just goes through the pedigree and verifies there are reasonable kinship coefficients for every first or second degree relation. This is likely going to be a one time run for each center to see if there are any issues.